src/operator/control_flow.cc

```diff
@@ -363,6 +363,393 @@ ForeachGradient(const nnvm::NodePtr& n, const std::vector<nnvm::NodeEntry>& ogra
+  }
+  template <typename T>
+  static void extract_by_loc(const std::vector<T> &array,
+                             const nnvm::Tuple<dim_t> input_locs,
+                             std::vector<T> *out) {
+    out->clear();
+    out->reserve(input_locs.ndim());
+    for (dim_t i : input_locs) {
+      out->push_back(array[i]);
+    }
+  }
+};
+
+static void WhileLoopComputeExCPU(const OpStatePtr& state_ptr,
+                                  const OpContext& ctx,
+                                  const std::vector<NDArray>& inputs,
+                                  const std::vector<OpReqType>& req,
+                                  const std::vector<NDArray>& outputs) {
+  // The argument `inputs' are loop_vars and other inputs
+  // loop_vars are stored in stored in `loop_vars_locs'
+  // The argument `outputs' are output and new_loop_vars
+  // [0: num_out_data) are outputs at each step.
+  // [num_out_data: ) are new_loop_vars
+  WhileLoopState &state = state_ptr.get_state<WhileLoopState>();
+  const WhileLoopParam& params = state.params;
+  // a helper function, converting std::vector<NDArray> to std::vector<NDArray*>
+  const auto to_ptr_vec = [](std::vector<NDArray> &in, std::vector<NDArray*> *out) {
+    out->clear();
+    out->reserve(in.size());
+    std::transform(std::begin(in), std::end(in), std::back_inserter(*out), [](NDArray &a) {return &a;});
+  };
+  // sanity checks
+  CHECK_EQ(inputs.size() + 2U, (size_t) params.num_args);
+  CHECK_EQ(outputs.size(), (size_t) params.num_outputs);
+  CHECK_EQ(outputs.size(), req.size());
+  CHECK_EQ(inputs.size() + params.num_out_data, outputs.size());
+  for (size_t i = 0; i < (size_t) params.num_out_data; i++)
+    CHECK_EQ(params.max_iterations, outputs[i].shape()[0]);
+  for (const auto &arr : outputs)
+    CHECK_EQ(arr.storage_type(), kDefaultStorage) << "The for operator doesn't support the sparse format";
+  // construct inputs and outputs for cond
+  std::vector<NDArray> cond_inputs, cond_outputs = {NDArray()};
+  WhileLoopState::extract_by_loc(inputs, params.cond_input_locs, &cond_inputs);
+  std::vector<NDArray*> cond_input_ptr, cond_output_ptr;
+  to_ptr_vec(cond_inputs, &cond_input_ptr);
+  to_ptr_vec(cond_outputs, &cond_output_ptr);
+  // a helper function to check is cond is satisfied
+  const auto check_cond = [&cond_input_ptr, &cond_output_ptr, &state]() {
+    // RAII storage
+    struct CPUStorage {
+      uint8_t *data;
+      size_t size;
+      CPUStorage(size_t size): data(new uint8_t[size]), size(size) {
+      }
+      ~CPUStorage() {
+        delete[] data;
+      }
+    };
+    // TODO(Junru): should I reset cond_output_ptr everytime? should I turn off sometime like is_recording?
+    // how to handle the storage?
+    // the SyncCopyToCPU here might be problematic
+    state.cond_op->Forward(nullptr, cond_input_ptr, cond_output_ptr);
+    NDArray *scalar = cond_output_ptr[0];
+    CPUStorage storage(scalar->shape().Size());
```

Comment:
the size may not be correct. it should be `scalar->shape().Size() * elem_size`

Reply:
solved